File: cpp/anagram/anagram.cpp
#include "anagram.h"

namespace anagram
{
	anagram::anagram(const string& word)
	{
		_orig = boost::to_lower_copy(word);
		_sorted = _sort(_orig);
	}

	vector<string> anagram::matches(const vector<string> words)
	{
		vector<string> ret;

		for (string word : words)
		{
			if (_matches(word))
			{
				ret.push_back(word);
			}
		}
		return ret;
	}

	bool anagram::_matches(const string& word)
	{
		string cp = boost::to_lower_copy(word);

		return _sort(cp) == _sorted && _orig != cp;
	}

	const string anagram::_sort(const string& word)
	{
		string cp = word;

		sort(cp.begin(), cp.end());
		return cp;
	}
}

File: java/anagram/src/main/java/Anagram.java
import java.util.Arrays;
import java.util.Collection;
import java.util.List;
import java.util.stream.Collectors;

public class Anagram {
    private String word;
    private String sortedWord;

    Anagram(String word) {
        this.word = word.toLowerCase();
        this.sortedWord = sort(this.word);
    }

    private static String sort(String orig) {
        char[] arr = orig.toCharArray();
        Arrays.sort(arr);
        return new String(arr);
    }

    private boolean isMatch(String anagram) {
        anagram = anagram.toLowerCase();
        return sort(anagram).equals(sortedWord) && !anagram.equals(word);
    }

    List<String> match(Collection<String> words) {
        return words.stream()
                .filter(this::isMatch)
                .collect(Collectors.toList());
    }
}
